![tracker](https://us-central1-vertex-ai-mlops-369716.cloudfunctions.net/pixel-tracking?path=statmike%2Fvertex-ai-mlops%2Fdata%2Bai%2Fbq-ai-functions%2Ffunctions%2Fai_search&file=ai_search.ipynb)
<!--- header table --->
<table align="left">
<tr>
  <td style="text-align: center">
    <a href="https://github.com/statmike/vertex-ai-mlops/blob/main/data%2Bai/bq-ai-functions/functions/ai_search/ai_search.ipynb">
      <img width="32px" src="https://www.svgrepo.com/download/217753/github.svg" alt="GitHub logo">
      <br>View on<br>GitHub
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://colab.research.google.com/github/statmike/vertex-ai-mlops/blob/main/data%2Bai/bq-ai-functions/functions/ai_search/ai_search.ipynb">
      <img width="32px" src="https://www.gstatic.com/pantheon/images/bigquery/welcome_page/colab-logo.svg" alt="Google Colaboratory logo">
      <br>Run in<br>Colab
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/vertex-ai/colab/import/https%3A%2F%2Fraw.githubusercontent.com%2Fstatmike%2Fvertex-ai-mlops%2Fmain%2Fdata%252Bai%2Fbq-ai-functions%2Ffunctions%2Fai_search%2Fai_search.ipynb">
      <img width="32px" src="https://lh3.googleusercontent.com/JmcxdQi-qOpctIvWKgPtrzZdJJK-J3sWE1RsfjZNwshCFgE_9fULcNpuXYTilIR2hjwN" alt="Google Cloud Colab Enterprise logo">
      <br>Run in<br>Colab Enterprise
    </a>
  </td>      
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/bigquery/import?url=https://github.com/statmike/vertex-ai-mlops/blob/main/data%2Bai/bq-ai-functions/functions/ai_search/ai_search.ipynb">
      <img width="32px" src="https://www.gstatic.com/images/branding/gcpiconscolors/bigquery/v1/32px.svg" alt="BigQuery logo">
      <br>Open in<br>BigQuery Studio
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/vertex-ai/workbench/deploy-notebook?download_url=https://raw.githubusercontent.com/statmike/vertex-ai-mlops/main/data%2Bai/bq-ai-functions/functions/ai_search/ai_search.ipynb">
      <img width="32px" src="https://www.gstatic.com/images/branding/gcpiconscolors/vertexai/v1/32px.svg" alt="Vertex AI logo">
      <br>Open in<br>Vertex AI Workbench
    </a>
  </td>
</tr>
<tr>
  <td colspan="5" style="text-align: right">
    <b>Share This On: </b> 
    <a href="https://www.linkedin.com/sharing/share-offsite/?url=https://github.com/statmike/vertex-ai-mlops/blob/main/data%252Bai/bq-ai-functions/functions/ai_search/ai_search.ipynb"><img src="https://upload.wikimedia.org/wikipedia/commons/8/81/LinkedIn_icon.svg" alt="Linkedin Logo" width="20px"></a> 
    <a href="https://reddit.com/submit?url=https://github.com/statmike/vertex-ai-mlops/blob/main/data%252Bai/bq-ai-functions/functions/ai_search/ai_search.ipynb"><img src="https://redditinc.com/hubfs/Reddit%20Inc/Brand/Reddit_Logo.png" alt="Reddit Logo" width="20px"></a> 
    <a href="https://bsky.app/intent/compose?text=https://github.com/statmike/vertex-ai-mlops/blob/main/data%252Bai/bq-ai-functions/functions/ai_search/ai_search.ipynb"><img src="https://upload.wikimedia.org/wikipedia/commons/7/7a/Bluesky_Logo.svg" alt="BlueSky Logo" width="20px"></a> 
    <a href="https://twitter.com/intent/tweet?url=https://github.com/statmike/vertex-ai-mlops/blob/main/data%252Bai/bq-ai-functions/functions/ai_search/ai_search.ipynb"><img src="https://upload.wikimedia.org/wikipedia/commons/5/5a/X_icon_2.svg" alt="X (Twitter) Logo" width="20px"></a> 
  </td>
</tr>
<tr>
  <td colspan="5" style="text-align: right">
    <b>Connect With Author On: </b> 
    <a href="https://www.linkedin.com/in/statmike"><img src="https://upload.wikimedia.org/wikipedia/commons/8/81/LinkedIn_icon.svg" alt="Linkedin Logo" width="20px"></a>
    <a href="https://www.github.com/statmike"><img src="https://www.svgrepo.com/download/217753/github.svg" alt="GitHub Logo" width="20px"></a> 
    <a href="https://www.youtube.com/@statmike-channel"><img src="https://upload.wikimedia.org/wikipedia/commons/f/fd/YouTube_full-color_icon_%282024%29.svg" alt="YouTube Logo" width="20px"></a>
    <a href="https://bsky.app/profile/statmike.bsky.social"><img src="https://upload.wikimedia.org/wikipedia/commons/7/7a/Bluesky_Logo.svg" alt="BlueSky Logo" width="20px"></a> 
    <a href="https://x.com/statmike"><img src="https://upload.wikimedia.org/wikipedia/commons/5/5a/X_icon_2.svg" alt="X (Twitter) Logo" width="20px"></a>
  </td>
</tr>
<tr>
  <td colspan="5" style="text-align: right">
    <a href="https://raw.githubusercontent.com/statmike/vertex-ai-mlops/main/data%2Bai/bq-ai-functions/functions/ai_search/ai_search.ipynb"><img src="https://www.svgrepo.com/download/5445/download-button.svg" alt="Download icon" width="20px"></a> <a href="https://raw.githubusercontent.com/statmike/vertex-ai-mlops/main/data%2Bai/bq-ai-functions/functions/ai_search/ai_search.ipynb">Download Notebook</a> <i>(right-click and "Save As")</i>
  </td>
</tr>
</table>

# AI.SEARCH — BigQuery AI Functions

`AI.SEARCH` is a simplified search function for tables with autonomous embedding generation enabled. It embeds the query at runtime and searches the table — no manual embedding step needed. The `mode` argument picks the retrieval strategy: semantic (vector) search, lexical (keyword) search fused with semantic search, or an automatic choice.

**Status:** `AI.SEARCH` is **GA**. Only the `mode` argument is **Preview** — hybrid search shipped 2026-06-25, was temporarily disabled 2026-07-09, and was restored 2026-08-03 (verified 2026-09-01). There is no separate `HYBRID_SEARCH` function; hybrid retrieval ships as `mode` here and as `lexical_search_columns` on [`VECTOR_SEARCH`](../vector_search/).

**When to use it:**
- You want the simplest possible semantic search (no embedding management)
- Your base table has autonomous embedding generation enabled
- You need single-query semantic search with a string literal
- You want hybrid (semantic + keyword) retrieval without assembling the lexical half yourself

**Alternatives:**
- [`VECTOR_SEARCH`](../vector_search/) — More control: batch search, custom embeddings, manual embedding management, and hybrid search over lexical columns you name explicitly
- [`AI.EMBED`](../ai_embed/) — Create individual embeddings for custom search logic
- [`AI.SIMILARITY`](../ai_similarity/) — Compare two specific inputs directly

**Featured in:** [Semantic Search System](../../workflows/semantic_search/) | [Catalog Search](../../workflows/catalog_search/)

**References:** [Full syntax reference](../../RESOURCES.md) | [Official documentation](https://cloud.google.com/bigquery/docs/reference/standard-sql/bigqueryml-syntax-ai-search) | [Setup guide](../../setup/)

---
## Setup

Set your project and location, authenticate, and create a temporary dataset for this notebook.

> This function requires a connection. The cells below create them if they don't exist. See the [Setup Reference](../../setup/) for details.

In [1]:
PROJECT_ID = 'statmike-mlops-349915'  # <-- Replace with your project ID
LOCATION = 'US'  # BigQuery dataset location
DATASET_ID = 'bq_ai_functions'  # Shared dataset across all notebooks
CONNECTION_ID = 'bq_ai_functions'  # Shared connection

### Environment

> **Already set up the project environment?** The cell below is a no-op — packages are already in your kernel. See the [Setup Reference](../../setup/) for details.
>
> **Running standalone** (Colab, Colab Enterprise, Vertex AI Workbench)? The cell below installs required packages into your current kernel.

In [2]:
import subprocess, sys, shutil

def install(*packages):
    """Install packages using uv (fast) with pip fallback."""
    uv = shutil.which('uv')
    if uv:
        subprocess.check_call([uv, 'pip', 'install', '-q', '--python', sys.executable, *packages])
    else:
        subprocess.check_call([sys.executable, '-m', 'pip', 'install', '-q', '--upgrade', *packages])

install('google-cloud-bigquery', 'db-dtypes', 'bigquery-magics', 'tqdm', 'bigframes')

In [3]:
# Authenticate (Colab only — other environments use Application Default Credentials)
try:
    from google.colab import auth
    auth.authenticate_user()
except ImportError:
    pass  # Not in Colab — ADC is used automatically

In [4]:
from google.cloud import bigquery
import pandas as pd

client = bigquery.Client(project=PROJECT_ID)
pd.set_option('display.max_colwidth', None)

# Create the shared dataset (idempotent)
dataset_ref = bigquery.DatasetReference(PROJECT_ID, DATASET_ID)
dataset = bigquery.Dataset(dataset_ref)
dataset.location = LOCATION
client.create_dataset(dataset, exists_ok=True)
print(f'Dataset {PROJECT_ID}.{DATASET_ID} ready')

# Register %%bigquery cell magic (auto-loaded in Colab, needed elsewhere)
%load_ext bigquery_magics

Dataset statmike-mlops-349915.bq_ai_functions ready


In [5]:
import subprocess as _sp, json as _json

# Create connection (idempotent)
_sp.run(['bq', 'mk', '--connection', '--location', LOCATION,
         '--connection_type', 'CLOUD_RESOURCE',
         '--project_id', PROJECT_ID, CONNECTION_ID],
        capture_output=True, text=True)

# Get service account and grant Vertex AI User role
r = _sp.run(['bq', 'show', '--connection', '--format=json',
             '--project_id', PROJECT_ID, '--location', LOCATION, CONNECTION_ID],
            capture_output=True, text=True, check=True)
sa = _json.loads(r.stdout)['cloudResource']['serviceAccountId']
_sp.run(['gcloud', 'projects', 'add-iam-policy-binding', PROJECT_ID,
         f'--member=serviceAccount:{sa}', '--role=roles/aiplatform.user', '--quiet'],
        capture_output=True, text=True)
print(f'Connection {CONNECTION_ID} ready (SA: {sa})')

Connection bq_ai_functions ready (SA: bqcx-1026793852137-hx0g@gcp-sa-bigquery-condel.iam.gserviceaccount.com)


---
## Examples — SQL

Progressive examples from simplest to most advanced. Each cell adds one new concept.

### Setup: Create a table with autonomous embedding generation

AI.SEARCH requires a base table with autonomous embedding generation enabled.
This is configured with a `GENERATED ALWAYS AS` column that calls `AI.EMBED`, with `OPTIONS(asynchronous = TRUE)`. BigQuery then automatically generates and maintains embeddings when data is inserted or updated. The same column definition works in `ALTER TABLE ... ADD COLUMN`, so an existing table can be upgraded in place.

**Note on task_type:** When you configure autonomous embedding generation, BigQuery handles the `task_type` automatically — it uses `RETRIEVAL_DOCUMENT` when indexing your table data and `RETRIEVAL_QUERY` when embedding search queries at runtime. This is the correct **asymmetric** pattern for retrieval. See the [`AI.EMBED`](../ai_embed/) notebook for the full list of task types.

**Limitations:**
- One automatically generated embedding column per table.
- The column must be `STRUCT<result ARRAY<FLOAT64>, status STRING>`; declaring it as a plain `ARRAY<FLOAT64>` fails with `Unsupported generated column expression`.
- `OPTIONS (asynchronous = TRUE)` is required — without it both `CREATE TABLE` and `ALTER TABLE ... ADD COLUMN` fail with `Generated embedding column requires asynchronous option to be true`.
- No column-level security policies (policy tags) on such tables, and no partitioned vector index.
- A vector index on the generated column is not supported at all — see example 6.

In [6]:
# Create table with autonomous embedding generation
query = f'''
CREATE OR REPLACE TABLE `{PROJECT_ID}.{DATASET_ID}.ai_search_knowledge_base` (
  id INT64,
  title STRING,
  content STRING,
  content_embedding STRUCT<result ARRAY<FLOAT64>, status STRING>
    GENERATED ALWAYS AS (
      AI.EMBED(content,
        connection_id => '{PROJECT_ID}.{LOCATION}.{CONNECTION_ID}',
        endpoint => 'text-embedding-005')
    ) STORED OPTIONS (asynchronous = TRUE)
)
'''
client.query(query).result()
print('Table created with autonomous embedding generation')

# Insert data — embeddings are generated automatically in the background
query = f'''
INSERT INTO `{PROJECT_ID}.{DATASET_ID}.ai_search_knowledge_base` (id, title, content)
VALUES
  (1, 'BigQuery', 'BigQuery is a serverless enterprise data warehouse for analytics at any scale.'),
  (2, 'Cloud Functions', 'Cloud Functions is a serverless execution environment for building event-driven applications.'),
  (3, 'Cloud Storage', 'Cloud Storage is a managed service for storing unstructured data of any size.'),
  (4, 'Kubernetes Engine', 'Google Kubernetes Engine provides a managed environment for deploying containerized applications.'),
  (5, 'Pub/Sub', 'Pub/Sub is an asynchronous messaging service that decouples services that produce events from services that process events.')
'''
client.query(query).result()
print('Data inserted — embeddings generating asynchronously')

# Wait for embeddings to be generated
import time
for attempt in range(30):
    df = client.query(f'''
        SELECT
          COUNT(*) AS total,
          COUNTIF(content_embedding IS NOT NULL AND content_embedding.status = '') AS ready
        FROM `{PROJECT_ID}.{DATASET_ID}.ai_search_knowledge_base`
    ''').to_dataframe()
    total, ready = int(df.iloc[0]['total']), int(df.iloc[0]['ready'])
    if ready == total:
        print(f'All {total} embeddings ready')
        break
    print(f'  Waiting... {ready}/{total} embeddings ready')
    time.sleep(10)
else:
    print(f'Warning: only {ready}/{total} embeddings ready after 5 minutes')

Table created with autonomous embedding generation
Data inserted — embeddings generating asynchronously
  Waiting... 0/5 embeddings ready
  Waiting... 0/5 embeddings ready
  Waiting... 0/5 embeddings ready
All 5 embeddings ready


### 1. Basic semantic search

AI.SEARCH takes the table, the column to search, and a query string. It embeds the query automatically and returns nearest matches.

In [7]:
query = f'''
SELECT base.title, base.content, distance
FROM AI.SEARCH(
  TABLE `{PROJECT_ID}.{DATASET_ID}.ai_search_knowledge_base`,
  'content',
  'serverless compute for running code'
)
'''
client.query(query).to_dataframe()

,title,content,distance
0,Cloud Functions,Cloud Functions is a serverless execution environment for building event-driven applications.,0.760961
1,Kubernetes Engine,Google Kubernetes Engine provides a managed environment for deploying containerized applications.,0.933319
2,BigQuery,BigQuery is a serverless enterprise data warehouse for analytics at any scale.,0.939442
3,Pub/Sub,Pub/Sub is an asynchronous messaging service that decouples services that produce events from services that process events.,0.967402
4,Cloud Storage,Cloud Storage is a managed service for storing unstructured data of any size.,0.968862


### 2. Limiting results with top_k

In [8]:
query = f'''
SELECT base.title, base.content, distance
FROM AI.SEARCH(
  TABLE `{PROJECT_ID}.{DATASET_ID}.ai_search_knowledge_base`,
  'content',
  'data storage and analytics',
  top_k => 2
)
'''
client.query(query).to_dataframe()

,title,content,distance
0,Cloud Storage,Cloud Storage is a managed service for storing unstructured data of any size.,0.774771
1,BigQuery,BigQuery is a serverless enterprise data warehouse for analytics at any scale.,0.875971


### 3. Changing distance type

In [9]:
query = f'''
SELECT base.title, base.content, distance
FROM AI.SEARCH(
  TABLE `{PROJECT_ID}.{DATASET_ID}.ai_search_knowledge_base`,
  'content',
  'messaging and events',
  top_k => 3,
  distance_type => 'COSINE'
)
'''
client.query(query).to_dataframe()

,title,content,distance
0,Pub/Sub,Pub/Sub is an asynchronous messaging service that decouples services that produce events from services that process events.,0.419300
1,Cloud Functions,Cloud Functions is a serverless execution environment for building event-driven applications.,0.483855
2,Cloud Storage,Cloud Storage is a managed service for storing unstructured data of any size.,0.620217


### 4. Hybrid search — `mode => 'HYBRID'` (Preview)

`mode` selects the retrieval strategy: `'VECTOR'` (semantic only), `'HYBRID'` (semantic fused with a lexical keyword search), or `'AUTO'` (the default). Hybrid aims at the exact tokens embeddings blur — SKUs, error codes, proper nouns — but it does not reach them by overriding the semantic leg. It fuses two ranked lists, so it changes *which* rows come back only when the two lists disagree, and how far down the corpus a keyword hit can reach is bounded by `top_k` twice over — once by the pool of rows the lexical leg is given, once by the fused score. Example 6 derives both gates and the reach they buy.

- `AI.SEARCH` takes **no** lexical arguments. The lexical half searches `column_to_search` with the same query string — there is no way to hand it an exact token of its own. Reach for [`VECTOR_SEARCH`](../vector_search/) when the keyword query has to run against *other* columns or carry its own text — it accepts explicit `lexical_search_columns` and `lexical_search_query_value`.
- No vector index is required. `mode => 'HYBRID'` runs on this five-row table with no index of any kind.
- `distance` changes meaning in hybrid mode. It stops being a distance — example 6 takes that apart.

In [10]:
query = f'''
SELECT base.title, base.content, distance
FROM AI.SEARCH(
  TABLE `{PROJECT_ID}.{DATASET_ID}.ai_search_knowledge_base`,
  'content',
  'serverless messaging between services',
  top_k => 3,
  mode => 'HYBRID'
)
'''
client.query(query).to_dataframe()

,title,content,distance
0,Pub/Sub,Pub/Sub is an asynchronous messaging service that decouples services that produce events from services that process events.,0.967478
1,Cloud Functions,Cloud Functions is a serverless execution environment for building event-driven applications.,0.967998
2,BigQuery,BigQuery is a serverless enterprise data warehouse for analytics at any scale.,0.968502


**Did the lexical half change the answer?** That is a claim about output, so measure it rather than assert it. The cell below runs the same query text in both modes and prints the set difference.

In [11]:
# Semantic-only vs hybrid on the same query text — compare the returned rows, not the scores
search_text = 'serverless messaging between services'

def top_titles(mode):
    query = f'''
    SELECT base.title
    FROM AI.SEARCH(
      TABLE `{PROJECT_ID}.{DATASET_ID}.ai_search_knowledge_base`,
      'content',
      '{search_text}',
      top_k => 3,
      mode => '{mode}'
    )
    ORDER BY distance
    '''
    return client.query(query).to_dataframe()['title'].tolist()

semantic, hybrid = top_titles('VECTOR'), top_titles('HYBRID')
print(f'semantic-only (VECTOR): {semantic}')
print(f'hybrid (HYBRID)       : {hybrid}')
print(f'Rows hybrid added that semantic-only missed: {[t for t in hybrid if t not in semantic] or "none"}')
print(f'Same rows, same order: {semantic == hybrid}')

semantic-only (VECTOR): ['Pub/Sub', 'Cloud Functions', 'BigQuery']
hybrid (HYBRID)       : ['Pub/Sub', 'Cloud Functions', 'BigQuery']
Rows hybrid added that semantic-only missed: none
Same rows, same order: True


**This corpus is too small to show a recall win, and that is worth understanding.** Five rows with `top_k => 3` means the semantic leg alone already returns 60% of the table, and every row here is an ordinary product description — there is no opaque token for the lexical leg to contribute that the embedding has not already found. Fusion has nothing to add, so it reorders at most.

Salting the corpus with error codes does not rescue the demo either: on documents this short a unique alphanumeric token dominates the embedding, so the semantic leg returns the row carrying the code first and hybrid has nothing left to recover (measured on a copy of this table, 2026-09-02).

Membership changes when two things line up: a token the embedding does not resolve, and a `top_k` deep enough for the fused score to reach the row carrying it. A rank-1 lexical match is worth `1/62` ≈ `0.0161`, which at `top_k => 3` reaches no further than semantic rank 6 — and the lexical leg is only handed the top `10 * top_k` rows to score, so even at `top_k => 64` nothing below semantic rank 640 can be recovered. Example 6 has the full reach table. For `VECTOR_SEARCH` there is a third lever: the exact token can be passed separately as `lexical_search_query_value` instead of being folded into the semantic query text.

See the [Catalog Search](../../workflows/catalog_search/) workflow for that head-to-head over a real product catalog, and the [`VECTOR_SEARCH`](../vector_search/) notebook for the same comparison with the underlying vector and lexical ranks printed side by side.

### 5. The default — `mode => 'AUTO'`, and the trap inside it

`AUTO` is the default, so examples 1-3 all ran in `AUTO` mode. The documented rule: `AUTO` performs a hybrid search **only if a hybrid index exists** — a vector index created with the `lexical_search_columns` option — and otherwise performs a semantic-only search.

**This is the failure mode worth memorizing.** `AUTO` never errors and never warns when it falls back. Ask for `AUTO` expecting hybrid, get a clean, plausible result set, and nothing in the output announces that the lexical half never ran. The only tell is the `distance` column: semantic distances spread out, fused hybrid scores cluster just under `0.97`.

This table cannot have a vector index at all (example 6 shows why), so `AUTO` resolves to semantic-only here and always will. The distances below are ordinary vector distances, identical to the `'VECTOR'` rows in example 6 and nothing like the fused scores in example 4. Set `mode` explicitly whenever hybrid behavior matters.

In [12]:
query = f'''
SELECT base.title, base.content, distance
FROM AI.SEARCH(
  TABLE `{PROJECT_ID}.{DATASET_ID}.ai_search_knowledge_base`,
  'content',
  'serverless messaging between services',
  top_k => 3,
  mode => 'AUTO'
)
'''
client.query(query).to_dataframe()

,title,content,distance
0,Pub/Sub,Pub/Sub is an asynchronous messaging service that decouples services that produce events from services that process events.,0.787263
1,Cloud Functions,Cloud Functions is a serverless execution environment for building event-driven applications.,0.802235
2,BigQuery,BigQuery is a serverless enterprise data warehouse for analytics at any scale.,0.992512


### 6. `mode => 'VECTOR'` — and what `distance` means in each mode

`'VECTOR'` forces a semantic-only search: the same work `AUTO` falls back to, but stated explicitly so the query never rides on `AUTO`'s index-sensitive default. On this table that default can never resolve to anything else — the reason is below the results. Running all three modes over one query text puts the difference on screen in a single result set.

Read the two scales separately:
- **`VECTOR`** returns a real distance in the space named by `distance_type` (`EUCLIDEAN` by default). Smaller is closer and an exact self-match is `0`.
- **`HYBRID`** returns a fused rank score. Smaller is still better, but it is not a distance: an exact self-match scores about `0.967`, not `0`, and the values are not comparable to `VECTOR` distances. Never put them on one axis or threshold them with one cutoff. `distance_type` has no effect in this mode: the fused score is built from the two rank lists, not from a metric space.

In [13]:
query = f'''
SELECT 'VECTOR' AS search_mode, base.title, distance
FROM AI.SEARCH(
  TABLE `{PROJECT_ID}.{DATASET_ID}.ai_search_knowledge_base`,
  'content', 'serverless messaging between services', top_k => 3, mode => 'VECTOR'
)
UNION ALL
SELECT 'AUTO', base.title, distance
FROM AI.SEARCH(
  TABLE `{PROJECT_ID}.{DATASET_ID}.ai_search_knowledge_base`,
  'content', 'serverless messaging between services', top_k => 3, mode => 'AUTO'
)
UNION ALL
SELECT 'HYBRID', base.title, distance
FROM AI.SEARCH(
  TABLE `{PROJECT_ID}.{DATASET_ID}.ai_search_knowledge_base`,
  'content', 'serverless messaging between services', top_k => 3, mode => 'HYBRID'
)
ORDER BY search_mode, distance
'''
client.query(query).to_dataframe()

,search_mode,title,distance
0,AUTO,Pub/Sub,0.787263
1,AUTO,Cloud Functions,0.802235
2,AUTO,BigQuery,0.992512
3,HYBRID,Pub/Sub,0.967478
4,HYBRID,Cloud Functions,0.967998
5,HYBRID,BigQuery,0.968502
6,VECTOR,Pub/Sub,0.787263
7,VECTOR,Cloud Functions,0.802235
8,VECTOR,BigQuery,0.992512


**What the hybrid score actually is.** BigQuery documents no fusion algorithm and returns no rank or score column. Reconstructed from observed output, hybrid `distance` is one minus a Reciprocal Rank Fusion score over the two ranked lists, with 1-based ranks:

```
distance = 1 - ( 1/(60 + rank_vector) + 1/(61 + rank_lexical) )
```

Read the two denominators carefully — they are not the same. The semantic leg uses a rank base of 60, the lexical leg 61. A row first semantically and second lexically returns `1 - (1/61 + 1/63)` = `0.9677335415040333`, and neither shared base reproduces that number: 60 on both legs would give `1 - (1/61 + 1/62)` = `0.9674775251189847`, and 61 on both legs `1 - (1/62 + 1/63)` = `0.9679979518689196`. The measured value falls *between* the two, which is the signature of mixed bases. Note that `0.9674775251189847` carries a second meaning that has nothing to do with base symmetry: under the real law it is the score of a row at rank 1 in *both* legs, which is the best score attainable. That is why hybrid scores sit just under `0.97` and a self-match is never `0`.

**The lexical leg ranks the entire candidate pool.** BigQuery hands the lexical half only the top `10 * top_k` rows by semantic rank, and that pool is the whole universe it sees. Inside it, true BM25 matches take lexical ranks 1 through *m*, and every remaining pooled row falls back to its semantic order behind them. Search with a lexical value that matches nothing and every returned row still carries a lexical term, with `rank_lexical` equal to `rank_vector`; promote one row to lexical rank 1 and the row that held rank 1 slides to 2 while everything below the promoted row is untouched. No pooled row is ever missing a term. Hybrid re-orders a single candidate set — it does not merge two lists in which a row can be present or absent. A row *outside* the pool is a different case entirely: it receives no lexical rank at all, however perfectly the token matches.

**How deep a keyword hit reaches.** Two gates decide it, and `top_k` sets both. The *pool* gate is the one above: only rows with `rank_vector <= 10 * top_k` are scored lexically. The *score* gate is the arithmetic — a rank-1 lexical match is worth `1/62` ≈ `0.0161`, so a matched row at semantic rank *R* scores `1/(60 + R) + 1/62`, and it has to beat the row holding the last slot, which sits at semantic rank `top_k` and, pushed down one position by the match, lexical rank `top_k + 1`. Effective reach is the smaller of the two, `min(score gate, 10 * top_k)`:

| `top_k` | 2 | 3 | 5 | 10 | 20 | 30 | 40 | 50 | 51 | 64 | 100 | 208 | 300 |
|---|---|---|---|---|---|---|---|---|---|---|---|---|---|
| deepest semantic rank retrievable | 3 | 6 | 10 | 23 | 56 | 110 | 212 | 468 | 510 | 640 | 1,000 | 2,080 | 3,000 |
| binding gate | score | score | score | score | score | score | score | score | pool | pool | pool | pool | pool |

The score gate binds below `top_k = 51`; from 51 up the pool binds and reach is simply `10 * top_k`. The sizing rule that falls out is a floor, not a recipe: **`top_k` of at least `R/10` is necessary to retrieve a row at semantic rank *R*, and sufficient only once the pool gate is the binding one.** While the score gate binds — below `top_k = 51`, so for targets less than a few hundred ranks deep — `R/10` is not enough: a target at semantic rank 100 needs `top_k = 29`, not 10, and one at rank 200 needs `top_k = 40`, not 20. Read the real requirement off the reach table above. The crossover is sharp, and it lands where the pool gate says it will — on a 3,000-row corpus a target at semantic rank 3,000 is absent at `top_k => 299` and returned at `top_k => 300`; on a 1,500-row corpus a target at rank 1,500 flips at `top_k => 150`. Those synthetic corpora sit below BigQuery's 5,000-row `CREATE VECTOR INDEX` floor and are brute-force scanned, but the gate is not an artifact of running unindexed: on a 7,275-row product catalog carrying an `ACTIVE`, fully covered hybrid `TREE_AH` index with `lexical_search_columns`, a target at semantic rank 2,072 still needs `top_k >= 208`, and `top_k => 64` returns nothing. An index neither causes the pool gate nor avoids it. Below roughly `top_k => 30` a keyword only nudges rows that were already near the top, which is exactly what example 4 shows at `top_k => 3`. Hybrid widens recall in proportion to `top_k`, never unboundedly. And when the identifier *is* the whole query and the answer has to be certain, a `WHERE` predicate cannot be outranked by a fusion score at all — and it has no pool.

This decomposition is reverse-engineered from observation, not documented by Google, and can change without notice — read it as an explanation of the shape of the numbers, never as an API contract. The [`VECTOR_SEARCH`](../vector_search/) notebook carries the full derivation with the underlying ranks visible.

**Why `AUTO` can never upgrade an `AI.SEARCH` base table.** A hybrid index is a vector index created with `lexical_search_columns`, where every lexical column also appears in a `STORING` clause and none of them is the index key:

```sql
CREATE VECTOR INDEX kb_hybrid_index
ON `PROJECT_ID.DATASET.some_table`(embedding)
STORING (content)
OPTIONS (index_type = 'IVF', distance_type = 'COSINE',
         lexical_search_columns = ['content']);
```

That index cannot exist on an `AI.SEARCH` base table. `AI.SEARCH` requires autonomous embedding generation, whose embedding column is a generated `STRUCT<result ARRAY<FLOAT64>, status STRING>`, and a vector index cannot be built on it:

```
CREATE VECTOR INDEX ... ON tbl(content_embedding)
-->  Vector index is not supported on column: 'content_embedding'
```

Nor on a field of it — `ON tbl(content_embedding.result)` returns `CREATE VECTOR INDEX does not yet support expressions to define index keys, only column name is supported` — and the generated column cannot be declared as a plain `ARRAY<FLOAT64>` to dodge the restriction. The DDL passes a dry run and fails at execution (verified 2026-09-01). Separately, a vector index only populates at 5,000+ rows and 10 MB, so this five-row demo table could not carry one in any case.

**The practical consequence:** on an `AI.SEARCH` base table today, `AUTO` is always semantic-only and `HYBRID` must be asked for by name. To get an *indexed* hybrid search, project `content_embedding.result` into a second table as a plain `ARRAY<FLOAT64>` column, build the hybrid index there, and query it with `VECTOR_SEARCH` — the [Catalog Search](../../workflows/catalog_search/) workflow builds that two-table design end to end.

---
## Examples — `%%bigquery` Magics

The same examples using IPython magic commands. Magics let you write SQL directly in notebook cells without Python string wrapping.

Key patterns:
- `%%bigquery` — run SQL, display results inline
- `%%bigquery df` — run SQL, capture results into a pandas DataFrame

### Semantic search with `%%bigquery`

In [14]:
%%bigquery --project {PROJECT_ID}

SELECT base.title, base.content, distance
FROM AI.SEARCH(
  TABLE `statmike-mlops-349915.bq_ai_functions.ai_search_knowledge_base`,
  'content',
  'container orchestration',
  top_k => 3
)

Query is running:   0%|          |

Downloading:   0%|          |

,title,content,distance
0,Kubernetes Engine,Google Kubernetes Engine provides a managed environment for deploying containerized applications.,0.834944
1,Pub/Sub,Pub/Sub is an asynchronous messaging service that decouples services that produce events from services that process events.,0.955050
2,Cloud Functions,Cloud Functions is a serverless execution environment for building event-driven applications.,0.963240


---
## Examples — BigFrames

`AI.SEARCH` has no direct BigFrames equivalent. Use `session.read_gbq_query()` to execute AI.SEARCH SQL from BigFrames.

In [15]:
import bigframes.pandas as bpd

bpd.options.bigquery.project = PROJECT_ID
bpd.options.bigquery.location = LOCATION

### Running AI.SEARCH via read_gbq_query

In [16]:
query = f"""
SELECT base.title, base.content, distance
FROM AI.SEARCH(
  TABLE `{PROJECT_ID}.{DATASET_ID}.ai_search_knowledge_base`,
  'content',
  'serverless compute',
  top_k => 3
)
"""
df = bpd.read_gbq_query(query)
df.to_pandas()

,title,content,distance
0,Cloud Functions,Cloud Functions is a serverless execution environment for building event-driven applications.,0.747659
1,BigQuery,BigQuery is a serverless enterprise data warehouse for analytics at any scale.,0.859217
2,Cloud Storage,Cloud Storage is a managed service for storing unstructured data of any size.,0.888576


---
## Cleanup

Drop resources created by this notebook. Shared resources (dataset, models, connection) are left for other notebooks.

In [17]:
client.delete_table(f'{PROJECT_ID}.{DATASET_ID}.ai_search_knowledge_base', not_found_ok=True)
print('Table ai_search_knowledge_base deleted')


Table ai_search_knowledge_base deleted


### Remove all shared resources

When finished with **all** notebooks, uncomment and run the cell below to delete the shared dataset and all tables, models, and other resources within it.

In [18]:
# client.delete_dataset(
#     f'{PROJECT_ID}.{DATASET_ID}',
#     delete_contents=True,
#     not_found_ok=True
# )
# print(f'Dataset {PROJECT_ID}.{DATASET_ID} deleted')